src/rust/en.mangafox/src/parser.rs

```diff
@@ -0,0 +1,362 @@
+		});
+	}
+
+	Ok(pages)
+}
+
+pub fn get_filtered_url(filters: Vec<Filter>, page: i32) -> String {
+	let mut is_searching = false;
+	let mut search_query = String::new();
+	let mut genre_query = String::new();
+	let mut url = String::from("https://fanfox.net");
+
+	for filter in filters {
+		match filter.kind {
+			FilterType::Title => {
+				if let Ok(filter_value) = filter.value.as_string() {
+					search_query.push_str("&name=");
+					search_query.push_str(urlencode(filter_value.read().to_lowercase()).as_str());
+					is_searching = true;
+				}
+			}
+			FilterType::Genre => {
+				if let Ok(filter_id) = filter.object.get("id").as_string() {
+					genre_query.push_str(filter_id.read().as_str());
+					is_searching = true;
+				}
+			}
+			FilterType::Select => {
+				if filter.name == "Language" {
+					search_query.push_str("&type=");
+					if filter.value.as_int().unwrap_or(-1) > 0 {
+						search_query
+							.push_str(&i32_to_string(filter.value.as_int().unwrap_or(-1) as i32));
+						is_searching = true;
+					}
+				}
+				if filter.name == "Rating" {
+					search_query.push_str("&rating_method=eq&rating=");
+					if filter.value.as_int().unwrap_or(-1) > 0 {
+						search_query
+							.push_str(&i32_to_string(filter.value.as_int().unwrap_or(-1) as i32));
+						is_searching = true;
+					}
+				}
+				if filter.name == "Completed" {
+					search_query.push_str("&st=");
+					if filter.value.as_int().unwrap_or(-1) > 0 {
+						search_query
+							.push_str(&i32_to_string(filter.value.as_int().unwrap_or(-1) as i32));
+						is_searching = true;
+					}
+				}
+			}
+			_ => continue,
+		}
+	}
+
+	if is_searching {
+		url.push_str("/search?page=");
+		url.push_str(&page.to_string());
+		url.push_str(&search_query);
+	} else {
+		url.push_str("/directory/");
+		url.push_str(&page.to_string());
+		url.push_str(".html?rating")
+	}
+	url
+}
+
+// pub fn parse_incoming_url(url: String) -> String {
+// 	// https://mangapill.com/manga/6290/one-piece-pirate-recipes
+// 	// https://mangapill.com/chapters/6290-10006000/one-piece-pirate-recipes-chapter-6
+
+// 	let split = url.as_str().split('/');
+// 	let vec = split.collect::<Vec<&str>>();
+// 	let mut manga_id = String::from("/manga/");
+
+// 	if url.contains("/chapters/") {
+// 		let split = vec[vec.len() - 2].split('-');
+// 		let ch_vec = split.collect::<Vec<&str>>();
+// 		manga_id.push_str(ch_vec[0]);
+// 	} else {
+// 		manga_id.push_str(vec[vec.len() - 2]);
+// 	}
+// 	manga_id.push('/');
+// 	manga_id.push_str(vec[vec.len() - 1]);
+// 	manga_id
+// }
+
+pub fn is_last_page(html: Node) -> bool {
+	let length = &html.select("div.pager-list-left a").array().len();
+	for (index, page) in html.select("div.pager-list-left a").array().enumerate() {
+		let page_node = page.as_node().expect("Failed to get page node");
+		let href = page_node.attr("href").read();
+		if index == length - 1 && href == "javascript:void(0)" {
+			return true;
+		}
+	}
+	false
+}
+
+// HELPER FUNCTIONS
```

Comment:
you don't need either of these functions, use uri_encode and alloc ToString

Reply:
removed helper functions, used uri::encode_uri from helper